File: Projects/Week 6 - Python/sentimental-readability/readability.py
from cs50 import get_string
import re

# Prompt the user for text input
text = get_string("Text: ")

# Initialize counters for letters, words, and sentences
letters = 0
words = 1  # Start at 1 to count the last word without a trailing space
sentences = 0

# Count letters, words, and sentences
for char in text:
    if char.isalpha():
        letters += 1
    elif char == ' ':
        words += 1
    elif char in ['.', '!', '?']:
        sentences += 1

# Calculate L and S as specified in the Coleman-Liau formula
L = (letters / words) * 100
S = (sentences / words) * 100

# Apply the Coleman-Liau formula
index = 0.0588 * L - 0.296 * S - 15.8

# Determine the grade level, rounded to nearest integer
grade = round(index)

# Print the appropriate grade level
if grade >= 16:
    print("Grade 16+")
elif grade < 1:
    print("Before Grade 1")
else:
    print(f"Grade {grade}")

File: Projects/Week 2 - Arrays/readability/readability.c
#include <cs50.h>
#include <ctype.h>
#include <math.h>
#include <stdio.h>
#include <string.h>

int count_letters(string text);
int count_words(string text);
int count_sentences(string text);

int main(void)
{
    // Get the input text from the user
    string text = get_string("Text: ");

    // Count the letters, words, and sentences in the text
    int letters = count_letters(text);
    int words = count_words(text);
    int sentences = count_sentences(text);

    // Calculate the average number of letters per 100 words
    float L = (float) letters / words * 100;

    // Calculate the average number of sentences per 100 words
    float S = (float) sentences / words * 100;

    // Compute the Coleman-Liau index
    float index = 0.0588 * L - 0.296 * S - 15.8;

    // Round the index to the nearest integer
    int grade = round(index);

    // Output the grade level
    if (grade < 1)
    {
        printf("Before Grade 1\n");
    }
    else if (grade >= 16)
    {
        printf("Grade 16+\n");
    }
    else
    {
        printf("Grade %i\n", grade);
    }
}

// Function to count the number of letters in the text
int count_letters(string text)
{
    int count = 0;
    for (int i = 0, n = strlen(text); i < n; i++)
    {
        if (isalpha(text[i]))
        {
            count++;
        }
    }
    return count;
}

// Function to count the number of words in the text
int count_words(string text)
{
    int count = 1; // Start count at 1 to account for the last word
    for (int i = 0, n = strlen(text); i < n; i++)
    {
        if (text[i] == ' ')
        {
            count++;
        }
    }
    return count;
}

// Function to count the number of sentences in the text
int count_sentences(string text)
{
    int count = 0;
    for (int i = 0, n = strlen(text); i < n; i++)
    {
        if (text[i] == '.' || text[i] == '!' || text[i] == '?')
        {
            count++;
        }
    }
    return count;
}
